File: Dante2/Module2.05/Zadanie5.02/main.c
#include <stdio.h>
#include <stdlib.h>

int **create_array_2d(int width, int height);
void display_array_2d(int **ptr, int width, int height);
void destroy_array_2d(int **ptr, int height);

int main()
{
    int error;

    int width, height;
    printf("Enter: ");

    error = scanf("%i %i", &width, &height);
    if (error != 2)
    {
        printf("Incorrect input");
        return 1;
    }

    if (width <= 0 || height <= 0)
    {
        printf("Incorrect input data");
        return 2;
    }

    int **array = create_array_2d(width, height);
    if (array == NULL)
    {
        printf("Failed to allocate memory");
        return 8;
    }

    printf("Enter: ");
    for (int i = 0; i < height; i++)
    {
        for (int j = 0; j < width; j++)
        {
            error = scanf("%i", &*(*(array + i) + j));
            if (error != 1)
            {
                printf("Incorrect input");
                destroy_array_2d(array, height);
                return 1;
            }
        }
    }

    display_array_2d(array, width, height);
    destroy_array_2d(array, height);

    return 0;
}

int **create_array_2d(int width, int height)
{
    if(width <= 0 || height <= 0)
    {
        return NULL;
    }

    int **array = (int **) malloc(sizeof(int *) * height);
    if(array == NULL)
    {
        return NULL;
    }

    for(int i = 0; i < height; i++)
    {
        *(array + i) = malloc(sizeof(int) * width);
        if(*(array + i) == NULL)
        {
            destroy_array_2d(array, height);
            return NULL;
        }
    }

    return array;
}

void display_array_2d(int **ptr, int width, int height)
{
    if (ptr != NULL)
    {
        if (width > 0 && height > 0)
        {
            for (int i = 0; i < height; i++)
            {
                for (int j = 0; j < width; j++)
                {
                    printf("%i ", *(*(ptr + i) + j));
                }
                printf("\n");
            }
        }
    }
}

void destroy_array_2d(int **ptr, int height)
{
    if (ptr == NULL)
    {
        return;
    }

    for (int i = 0; i < height; i++)
    {
        free(*(ptr + i));
    }
    free(ptr);
}

File: Dante2/Module2.05/Zadanie5.02/main.rs
use regex::Regex;

fn create_array_2d(width: i32, height: i32) -> Result<Vec<Vec<i32>>, String> {
    if width <= 0 || height <= 0 {
        return Err("Width and height must be positive integers.".to_string());
    }

    let array: Vec<Vec<i32>> = vec![vec![0; width as usize]; height as usize];

    Ok(array)
}

fn display_array_2d(array: & Vec<Vec<i32>>, width: i32, height: i32) {
    if width < 0 || height < 0 {
        return;
    }
    for i in 0..height {
        for j in 0..width {
            print!("{} ", array[i as usize][j as usize]);
        }
        println!();
    }
}

fn main() {
    let mut input: String = String::new();
    println!("Enter: ");
    std::io::stdin().read_line(&mut input).expect("Incorrect input");
    let re = Regex::new(r"(\d+)\s+(\d+)").expect("Incorrect REGEX");
    if let Some(caps) = re.captures(&input) {
        let width: i32 = caps[1].parse().expect("Incorrect width parse");
        let height: i32 = caps[2].parse().expect("Incorrect height parse");

        if width <= 0 || height <= 0 {
            panic!("Incorrect input data");
        }

        match create_array_2d(width, height) {
            Ok(mut array) => {
                println!("Enter: ");

                for i in 0..height {
                    for j in 0..width {
                        input.clear();
                        std::io::stdin().read_line(&mut input).expect("Incorrect input in array");
                        array[i as usize][j as usize] = input.trim().parse().expect("Incorrect parsing");
                    }
                }

                display_array_2d(&array, width, height);
            }
            Err(e) => eprintln!("Error: {}", e),
        }
    }
    else {
        panic!("Incorrect numbers");
    }
}
